File: backend/app/schemas/scenario.py
"""Pydantic schemas for scenario planning (what-if analysis)."""

from typing import Dict, List, Literal, Optional, Union

from pydantic import BaseModel, Field


class AddProjectPayload(BaseModel):
    id: Optional[str] = None
    name: str
    type: Optional[str] = None
    portfolio: Optional[str] = None
    sponsor: Optional[str] = None
    priority: Optional[str] = "Medium"
    start_date: str
    end_date: str
    est_hours: float = 0.0
    role_allocations: Dict[str, float] = Field(default_factory=dict)


class AddPersonPayload(BaseModel):
    name: str
    role_key: str
    role: Optional[str] = None
    team: Optional[str] = None
    vendor: Optional[str] = None
    classification: Optional[str] = None
    rate_per_hour: float = 0.0
    weekly_hrs_available: float = 40.0
    support_reserve_pct: float = 0.0


class AddProjectMod(BaseModel):
    type: Literal["add_project"]
    project: AddProjectPayload


class CancelProjectMod(BaseModel):
    type: Literal["cancel_project"]
    project_id: str


class ExcludePersonMod(BaseModel):
    type: Literal["exclude_person"]
    person_name: str


class AddPersonMod(BaseModel):
    type: Literal["add_person"]
    person: AddPersonPayload


class ShiftProjectMod(BaseModel):
    type: Literal["shift_project"]
    project_id: str
    new_start_date: Optional[str] = None
    new_end_date: Optional[str] = None


class ChangeAllocationMod(BaseModel):
    type: Literal["change_allocation"]
    project_id: str
    role_key: str
    allocation: float = Field(ge=0.0, le=1.0)


class ResizeProjectMod(BaseModel):
    type: Literal["resize_project"]
    project_id: str
    est_hours: float = Field(ge=0.0)


Modification = Union[
    AddProjectMod,
    CancelProjectMod,
    ExcludePersonMod,
    AddPersonMod,
    ShiftProjectMod,
    ChangeAllocationMod,
    ResizeProjectMod,
]


class ScenarioEvaluateRequest(BaseModel):
    name: Optional[str] = None
    modifications: List[Modification] = Field(default_factory=list)


class RoleUtilSnapshot(BaseModel):
    role_key: str
    supply_hrs_week: float
    demand_hrs_week: float
    utilization_pct: float
    status: str


class UtilizationSide(BaseModel):
    roles: Dict[str, RoleUtilSnapshot]


class ScenarioDelta(BaseModel):
    role_key: str
    baseline_pct: float
    scenario_pct: float
    delta_pct: float
    baseline_status: str
    scenario_status: str
    status_changed: bool


class ScenarioSummary(BaseModel):
    headline: str
    became_over: List[str]
    became_stretched: List[str]
    became_unstaffed: List[str]
    became_better: List[str]


class ScenarioEvaluateResponse(BaseModel):
    baseline: UtilizationSide
    scenario: UtilizationSide
    deltas: List[ScenarioDelta]
    summary: ScenarioSummary


class SchedulePortfolioRequest(BaseModel):
    max_util_pct: Optional[float] = Field(None, ge=0.1, le=2.0)
    horizon_weeks: int = Field(52, ge=4, le=156)
    exclude_ids: List[str] = Field(default_factory=list)
    modifications: List[Modification] = Field(default_factory=list)


class ScheduledProject(BaseModel):
    project_id: str
    project_name: str
    priority: str
    est_hours: float
    health: str
    suggested_start: Optional[str] = None
    suggested_end: Optional[str] = None
    duration_weeks: float
    wait_weeks: Optional[int] = None
    bottleneck_role: Optional[str] = None
    can_start_now: bool


class InFlightProject(BaseModel):
    project_id: str
    project_name: str
    priority: str
    est_hours: float
    health: str
    pct_complete: float
    start_date: Optional[str] = None
    end_date: Optional[str] = None


class SchedulePortfolioResponse(BaseModel):
    max_util_pct: float
    horizon_weeks: int
    in_flight: List[InFlightProject]
    projects: List[ScheduledProject]
    can_start_now_count: int
    waiting_count: int
    infeasible_count: int
    bottleneck_roles: Dict[str, int]

File: frontend/src/types/scenario.ts
import type { RoleStatus } from "./capacity";

// ---------------------------------------------------------------------------
// Modification payloads — discriminated union on `type`
// ---------------------------------------------------------------------------

export interface AddProjectPayload {
  id?: string;
  name: string;
  type?: string | null;
  portfolio?: string | null;
  sponsor?: string | null;
  priority?: string;
  start_date: string; // ISO YYYY-MM-DD
  end_date: string;
  est_hours: number;
  role_allocations: Record<string, number>;
}

export interface AddPersonPayload {
  name: string;
  role_key: string;
  role?: string | null;
  team?: string | null;
  vendor?: string | null;
  classification?: string | null;
  rate_per_hour?: number;
  weekly_hrs_available: number;
  support_reserve_pct?: number;
}

export type ScenarioModification =
  | { type: "add_project"; project: AddProjectPayload }
  | { type: "cancel_project"; project_id: string }
  | { type: "exclude_person"; person_name: string }
  | { type: "add_person"; person: AddPersonPayload }
  | { type: "shift_project"; project_id: string; new_start_date?: string; new_end_date?: string }
  | { type: "change_allocation"; project_id: string; role_key: string; allocation: number }
  | { type: "resize_project"; project_id: string; est_hours: number };

// ---------------------------------------------------------------------------
// Request / response
// ---------------------------------------------------------------------------

export interface ScenarioEvaluateRequest {
  name?: string;
  modifications: ScenarioModification[];
}

export interface RoleUtilSnapshot {
  role_key: string;
  supply_hrs_week: number;
  demand_hrs_week: number;
  utilization_pct: number;
  status: RoleStatus;
}

export interface UtilizationSide {
  roles: Record<string, RoleUtilSnapshot>;
}

export interface ScenarioDelta {
  role_key: string;
  baseline_pct: number;
  scenario_pct: number;
  delta_pct: number;
  baseline_status: RoleStatus;
  scenario_status: RoleStatus;
  status_changed: boolean;
}

export interface ScenarioSummary {
  headline: string;
  became_over: string[];
  became_stretched: string[];
  became_unstaffed: string[];
  became_better: string[];
}

export interface ScenarioEvaluateResponse {
  baseline: UtilizationSide;
  scenario: UtilizationSide;
  deltas: ScenarioDelta[];
  summary: ScenarioSummary;
}

// ---------------------------------------------------------------------------
// Auto-scheduler — project placement
// ---------------------------------------------------------------------------

export interface SchedulePortfolioRequest {
  max_util_pct?: number;
  horizon_weeks?: number;
  exclude_ids?: string[];
  modifications?: ScenarioModification[];
}

export interface ScheduledProject {
  project_id: string;
  project_name: string;
  priority: string;
  est_hours: number;
  health: string;
  suggested_start: string | null;
  suggested_end: string | null;
  duration_weeks: number;
  wait_weeks: number | null;
  bottleneck_role: string | null;
  can_start_now: boolean;
}

export interface InFlightProject {
  project_id: string;
  project_name: string;
  priority: string;
  est_hours: number;
  health: string;
  pct_complete: number;
  start_date: string | null;
  end_date: string | null;
}

export interface SchedulePortfolioResponse {
  max_util_pct: number;
  horizon_weeks: number;
  in_flight: InFlightProject[];
  projects: ScheduledProject[];
  can_start_now_count: number;
  waiting_count: number;
  infeasible_count: number;
  bottleneck_roles: Record<string, number>;
}
